gax-java/gax/src/test/java/com/google/api/gax/rpc/ApiClientHeaderProviderTest.java

```diff
@@ -138,4 +138,18 @@ public void testQuotaProjectHeader() {
     assertThat(provider.getHeaders().get(ApiClientHeaderProvider.QUOTA_PROJECT_ID_HEADER_KEY))
         .matches(quotaProjectHeaderValue);
   }
+
+  @Test
+  public void testApiVersionHeader() {
+    ApiClientHeaderProvider provider =
+        ApiClientHeaderProvider.newBuilder().setApiVersionToken("fake-version-string").build();
+    assertThat(provider.getHeaders().size()).isEqualTo(2);
+    assertThat(provider.getHeaders().get(ApiClientHeaderProvider.API_VERSION_HEADER_KEY))
+        .matches("fake-version-string");
+
+    ApiClientHeaderProvider emptyProvider =
+        ApiClientHeaderProvider.newBuilder().setApiVersionToken("").build();
+    assertThat(
+        emptyProvider.getHeaders().get(ApiClientHeaderProvider.API_VERSION_HEADER_KEY).isEmpty());
```

Comment:
My personal opinion is that I think this case should result in the ApiVersion header not being passed (similar to how it would behave if the service proto doesn't have this option). I know the client library can't reasonably do validation all values passed in to the ApiClientHeader, but empty string and null are cases which we know aren't valid ApiVersions.

The reason I have this case in mind is because it is something we check for inside the EndpointContext: https://github.com/googleapis/sdk-platform-java/blob/75352d197f1744435cca83ac788edea51de787f8/gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java#L220-L222

Reply:
Thanks for the context. 

The way I see it currently is that, this logic of validating api version should not be repeated if possible. One benefit is that in case of future changes to it, say additional logic to treat a group of versions in a way (totally hypothetical here), change can be centralized. 
Currently, this apiVersion information flows from:
proto --> via Parser --> model (Service) --> via composer --> output code (StubSettings) --> creates a ApiClientHeaderProvider instance.
`ApiClientHeaderProvider.java` does not need to know the details about how to validate api versions because it's already validated upstream.

WDYT?

Reply:
I think we might be on two different pages here. I don't intend for ApiClientHeaderProvider to try and validate the value that is set by the service team inside the proto file. If that value is not set by the team or is set as empty, then I trust that the generator has the correct logic to determine when the `setApiVersionToken()` code should or should not be generated.

My concern is that both ApiClientHeaderProvider and the ApiVersionToken setter is public. Users can create an ApiClientHeaderProvider separately and then pass it into the client via StubSettings (https://github.com/googleapis/sdk-platform-java/blob/849bb503b9e09310e8225c2997a5ceeb866dab31/gax-java/gax/src/main/java/com/google/api/gax/rpc/StubSettings.java#L442-L449). I believe it's possible given that we previously have suggested to users to set their own header provider values for certain workaround use cases.

I agree that's it's a extremely rare possibility that users are both creating a custom HeaderProvider and intentionally setting the ApiVersionToken to `""`, but because it's possible, I don't want to rule that possibility out. IMO, we should keep consistent with the default behavior done automatically for GAPICs (i.e. if it is set to be null or empty string, then it's not set in the header).

Granted this is a small nit pick that I think is a better UX. If you feel strongly otherwise, I don't want to continue to block on this small part. I think the other bits LGTM and I will do another pass before approving. Thanks for all the changes :)

Reply:
Followup discussions directed off-line. 
Merging this pr for now. May raise followup based on discussion result.